### ISSUE TITLE
FinalLocalVariable: false negative on pattern variables

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#10970]
Identified at https://github.com/checkstyle/checkstyle/issues/7504#issuecomment-974776332 ,

https://checkstyle.sourceforge.io/config_coding.html#FinalLocalVariable

````
$ cat TestClass.java
public class Test {
    public void run(String... arguments) {
        final Object o = 45;
        if (o instanceof String p) { // line 4
            System.out.println(p);
        }
        if (o instanceof final String p) { // line 7
            System.out.println(p);
        }
        if (o instanceof String p) { // line 10
            p = "rewrite";
            System.out.println(p);
        }
    }
}


$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="FinalLocalVariable">
</module>
    </module>
</module>

$ java -jar checkstyle-9.0.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I am expecting a violation on line 4 as line 7 shows that the pattern variables can be declared `final` and can be re-written as shown on line 11.

https://openjdk.java.net/jeps/394
> Lift the restriction that pattern variables are implicitly final, to reduce asymmetries between local variables and pattern variables.

**Comment by romani** [comment]
As we have special Check for this pattern variable , I think it is ok to have no violation.
We might need to mention explicitly final to be a target of Check.

**Comment by rnveach** [comment]
> lambda variable we also have special Check https://checkstyle.org/config_naming.html#LambdaParameterName

That is for naming, which I am not asking for in this issue.
Are you say we should have FinalPatternVariable check or we should not violate this at all?

> We might need to mention explicitly final to be a target of Check.

I am not sure what you are saying here. I am asking for a violation to printed if a pattern variable isn't re-assigned and is not final so the user can write the words final. You say its ok to have to no violation, but then I assume you mention that we want final to be the target of the check which I am saying too.

**Comment by romani** [comment]
>we should not violate this at all?

Yes, no violation at all

**Comment by rnveach** [comment]
Then what about a separate check like `FinalPatternVariable`?

**Comment by nmancus1** [comment]
> Then what about a separate check like `FinalPatternVariable`?

I think this is a good idea; perhaps even extending the check (or having another) that verifies that all pattern variables are declared as final. Pattern variable reassignment seems like a bad practice to me.

This check will require expression evaluation as discussed in https://github.com/checkstyle/checkstyle/issues/7504#issuecomment-974820323 to determine exact scope of pattern variable.

**Comment by romani** [comment]
I removed all my comments, I were confused.
I updated issue description.

I think we should update existing Check. Declared expectations in issue is valid.

@pbludov , @strkkk , please vote.

**Comment by strkkk** [comment]
it seems a bit weird to me to make a special check for such small case. I think it fits into existing check

**Comment by romani** [comment]
I approved issue, looks like majority agree to update existing Check

**Comment by Vyom-Yadav** [comment]
I am on it, will first implement the expression evaluator over here and then will solve #10969

**Comment by pbludov** [comment]
I think that  pattern variables should be similar to method parameters.

My opinion is:
* No need for `final` modifier for pattern variables.
* We need  a check to ensure that they are not changed, similar to [ParameterAssignment](https://checkstyle.org/config_coding.html#ParameterAssignment).

**Comment by romani** [comment]
@pbludov , do think that FinalLocalVariable should not place violation on code that in PR description ?

**Comment by pbludov** [comment]
> @pbludov , do think that FinalLocalVariable should not place violation on code that in PR description ?

`FinalLocalVariable` should process only regular variables, i.e.
```java
final String result;
```
Pattern variables are more like lambdas, their context is shorter, and naming requirements can be less strict.

**Comment by nrmancuso** [comment]
Ok, let's extend FinalLocalVariable with a new property, `validatePatternVariables`.

**Comment by mahfouz72** [comment]
following from https://github.com/checkstyle/checkstyle/issues/10970#issuecomment-986003441
@nrmancuso what about a check that violates the assignment of pattern variables?

**Comment by nrmancuso** [comment]
> following from [#10970 (comment)](https://github.com/checkstyle/checkstyle/issues/10970#issuecomment-986003441) @nrmancuso what about a check that violates the assignment of pattern variables?

Yes, please create a ticket for this

**Comment by Atharv3221** [comment]
I am working on this

**Comment by romani** [comment]
@Sahitya0805, can you help us?
look at closed PR, it has a lot to reuse, it was closed only because @Atharv3221  do not have time to finish due to other priorities

**Comment by sahitya0xsingh** [comment]
Sure, I'll take a look at the closed PR, reuse as much as possible, and continue the remaining work. Thanks!

**Comment by romani** [comment]
@sahitya0xsingh , please create new issue for new Check creation, and we can close this issue

**Comment by sahitya0xsingh** [comment]
@romani Thanks for the guidance. I have created the new issue for the dedicated `FinalPatternVariableCheck`: https://github.com/checkstyle/checkstyle/issues/21359

I’ll proceed with the implementation there.

**Comment by romani** [comment]
Ok will continue at https://github.com/checkstyle/checkstyle/issues/21359
